### ISSUE TITLE
Feature Request/Discussion - Expansion for function types

### ISSUE THREAD

**Issue by joncalhoun** [nsf/gocode#483]
When writing Go code it is fairly common to have a type like this:

```go
type HandlerFunc func(ResponseWriter, *Request)
```

Or more generally...

```go
type Xxxx func(T1, T2, T3) T4
```

Where `Xxxx` is the type name, and `T1`, `T2`, etc are all types.

What I would love to be able to do is to take that information and be able to create code completions out of them. For instance, if I typed something like `http.HandlerFunc(` and hit tab I would love to be able to turn that into a snippet in Atom like:

```
func($1 http.ResponseWriter, $2 *http.Request) {
  $3
}
```

Where `$1` through `$3` are all tab-stops in their snippets.

The goal here is to make it easier to create closures etc when writing some code like this:

```go
func demo(...) http.HandlerFunc {
  return func(w http.ResponseWriter, r *http.Request) {
    // ...
  }
}
```

I started [discussing the issue with the go-plus maintainers](https://github.com/joefitzgerald/go-plus/issues/703) and it sounds like this might require some additions to gocode as well as the go-plus plugin.

I know very little about `gocode` and how exactly it works, but I'm willing to try to dig into this to see if it is plausible if someone could help point me in the right direction. That would enable me to at least build a prototype and discuss if it makes sense to fully merge the feature into gocode.

Thanks!

**Comment by nsf** [comment]
What I can do on my side is that if there is a function argument in a function and it's defined via named type, I can dereference it and present the actual function type.

E.g. 
```go
```

**Comment by nsf** [comment]
(sorry for messed up comment, if you received it via email, I submitted it by accident earlier than I wanted to, edited the comment on github though)

**Comment by joncalhoun** [comment]
What you mentioned is already incorporated into `go-plus` I believe, but it is slightly different than what I would like to add. 

In the example you provided I need to be calling a function (`foo`) that accepts a `Foo` argument for the expansion to occur. 

What I would like to do is create expansions of type `Foo` into its underlying function type regardless of context. That way I could do something like:

```go
type DogFunc func(a int, b string) string

func demo() (DogFunc, error) {
  df := DogFunc(<hit tab or activate autocomplete here>
}
```

And then the written text of `DogFunc(` could be expanded into:

```go
func demo() (DogFunc, error) {
  df := func(a int, b string) string {

  }
}
```

In other words, when I'm creating an `http.HandlerFunc` closure, or otherwise creating an instance of a type that has a function as its underlying type, I'd like a way to expand that.

I could probably write a tool that scrapes all of my $GOPATH and finds every `type Xxxx func...` type and creates snippets for my editor (Atom), but the downside to this approach is that it doesn't take code changes into effect. Using something like gocode instead would mean that if `DogFunc` were changed then the autocompletion could be updated dynamically as well.

Does that make sense?

**Comment by nsf** [comment]
Yeah, actually I don't think I can help you here. Because what you're describing is editor's behaviour. And from gocode I can only influence what I return based on cursor position. As I said I can return a raw function type even if it's actually a named type. Maybe gocode does it already, I don't remember. But that's all there is. It's up to an editor to do any kind of substitution from here.

**Comment by joncalhoun** [comment]
I realize the substitution is up to the editor plugin, but I was basically wondering if there was a way with gocode to get the underlying type of a type. Eg if I am typing:

```go
type Foo func(a int)

func main() {
  Foo<I request info about this location>
}
```

Would it be possibel to get back information about the underlying type being `func(a int)`?

Your last comment I think suggests that you aren't sure if gocode does this now or not.

**Comment by joncalhoun** [comment]
Closing this as it's become clear this feature is definitely plausible without any gocode changes.
